KAFKA-12639: exit upon expired timer to prevent tight looping

philipnee opened the pull request:
https://issues.apache.org/jira/browse/KAFKA-12639

In AbstractCoordinator#joinGroupIfNeeded - joinGroup request will be retried without proper backoff, due to the expired timer. This is an uncommon scenario and possibly only appears during the testing, but I think it makes sense to enforce the client to drive the join group via poll.

philipnee committed:
exit upon expired timer to prevent tight looping

philipnee committed:
exit upon expired timer to prevent tight looping

philipnee commented:
@guozhangwang - would you have time to review this 🥺 ?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -501,13 +501,16 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 else if (!future.isRetriable())
                     throw exception;
 
+                if (timer.isExpired()) {
```
Could you add a couple comments here explaining why we check the timer again here in addition to in line 452 above? Maybe something like this:

```
We check the timer again after calling poll with the timer since it's possible that even after the timer has elapsed, the next client.poll(timer) would immediately return an error response which would cause us to not exiting the while loop.
```

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -501,13 +501,16 @@ boolean joinGroupIfNeeded(final Timer timer) {
-                    exception instanceof MemberIdRequiredException)
+                        exception instanceof IllegalGenerationException ||
+                        exception instanceof RebalanceInProgressException ||
+                        exception instanceof MemberIdRequiredException)
```
Should we actually do the timer check before this? Since otherwise if the exception from the immediately returned responses is any of those four, we would still `continue` and skip the check below.

More concretely I think we can just move the remaining logic inside the `if` call:

```
if (!future.isRetriable()) {
    throw ..
} else {
    if (timer.isExpired()) {
        return false;
    } else if (exception instance of..) {
        continue;
    } else {
        timer.sleep(..)
    }
}
```

philipnee replied:
agreed. A comment here, retriableException check should happen after the `instanceOf` checks, because I think we actually want to retry upon these (according to the logic).

guozhangwang reviewed (commented):
Thanks @philipnee . I left some comments.

Also I think we should add unit test with exactly the mocking clients  to 1) return a non-retriable exception, and check that we throw immediately (if we already have it in the test, then we can skip), 2) return one of the four exceptions, and check that we never sleep backoffs, 3) return other retriable exception, and check that we would still the remaining timer; and in case 2/3), we check that if timer has elapsed, we would also return false immediately.

philipnee committed:
Added tests according to the comments

philipnee committed:
Added tests according to the comments

philipnee commented:
Thanks @guozhangwang for the feedback - Added some tests there to cover the untesed cases.  I still have a quick question around this block, is it intentional to continue w/o sleep on the backoff timer? (quoting the original code)

```
if (exception instanceof UnknownMemberIdException ||
                    exception instanceof IllegalGenerationException ||
                    exception instanceof RebalanceInProgressException ||
                    exception instanceof MemberIdRequiredException)
                    continue;
                else if (!future.isRetriable())
                    throw exception;

                timer.sleep(rebalanceConfig.retryBackoffMs);
```

guozhangwang commented:
> is it intentional to continue w/o sleep on the backoff timer?

Yes that's intentional. For those four exceptions, we'd like to send the follow-up request right away since the broker is waiting for those join-group request. But the question is, when the timer has already elapsed, should we honor that or should we ignore but always try to complete this mid-stage.

Since in the new protocol we would no longer have such mid-stages during a prepare_rebalance phase (cc @dajac to chime in if you feel different), I would suggest we respect the timer still for now to have a stronger `poll(timer)` timing guarantees.

philipnee commented:
Thanks, @guozhangwang, that's my understanding as well.

philipnee committed:
re-trigger test

abd

philipnee committed:
re-trigger test

abd

philipnee committed:
Moving the timer below the specific exceptions to prevent breaking changes.

philipnee committed:
Moving the timer below the specific exceptions to prevent breaking changes.

philipnee committed:
Merge branch 'kafka-12639_exit_upon_expired_timer' of github.com:philipnee/kafka into kafka-12639_exit_upon_expired_timer

philipnee committed:
Merge branch 'apache:trunk' into kafka-12639_exit_upon_expired_timer

philipnee committed:
retrigger test

retrigger test

guozhangwang commented:
@philipnee is this the final version of this PR? Seems we are still honoring the four exceptions indicating the mid-stage of a rebalance more than the elapsed timer here?

philipnee commented:
Hey @guozhangwang it's WIP - I think moving the timer check before the exception handling block (that 4 exceptions), kind of breaks a bunch of tests, as most tests are expecting the complete within a single poll. I'm looking into these breakage actually. sorry about the confusion.

guozhangwang commented:
Oh got it, thanks! All good :) Please let me know when it's ready for a final look.

philipnee commented:
Although, are we ok with handling these 4 exceptions the same way as before? I know you previously mentioned that it might be better off to make the rules more consistent, and I kind of agree with it.

philipnee committed:
wip

philipnee committed:
Make sure joinGroupIfNeeded is called twice

- First call will exit upon timeout
- Second call should send a proper request before exiting.

philipnee committed:
re-trigger test

re trigger-test

guozhangwang commented:
Yeah I think it's okay to make the rule consistent, i.e. to honor the timeout even under those four exceptions: if the timer has elapsed, then we should well return from the loop in

```
client.poll(future, timer);
            if (!future.isDone()) {
                // we ran out of time
                return false;
            }
```

even if the response yet to be returned would contain any of these four exceptions. So I think we should still obey this rule, i.e. even if a response has been returned and we know it's going to be one of these four exceptions, if the timer has elapsed, we still exit the loop.

philipnee commented:
Hmm, strangely, this branch seems to trigger a bunch of initializing error failures. And I can't seem to reproduce them locally...

philipnee committed:
clean up

philipnee committed:
Need to be more careful at handling timeout.

philipnee committed:
fix indentation

philipnee committed:
1

philipnee committed:
retrigger test

philipnee commented:
The failures seem irrelevant to the change here: i.e. they dont' show up in both rounds.
```
Build / JDK 11 and Scala 2.13 / testDynamicListenerConnectionCreationRateQuota() – kafka.network.DynamicConnectionQuotaTest
41s
Build / JDK 17 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
2m 0s
Build / JDK 8 and Scala 2.12 / [1] Type=ZK, Name=testRegisterZkBrokerInKraft, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.server.KafkaServerKRaftRegistrationTest
7s
Build / JDK 8 and Scala 2.12 / [1] Type=ZK, Name=testRegisterZkBrokerInKraft, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.server.KafkaServerKRaftRegistrationTest
12s
```

philipnee committed:
nit

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -500,14 +500,22 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     requestRejoin(shortReason, fullReason);
                 }
 
+                // continue to retry as long as the timer hasn't expired
```
Could we simplify this multi-if logic as:

```
if (!future.isRetriable()) { throw }
else {
    if (timer.isExpired() { return false }
    else if (exception instance of.. ) { continue}
    else {timer.sleep(..)}
}
```

Also could we add a comment on top clarifying that the order of precedence are deliberated in this order and future changes should pay attention to not change it unnecessarily.

philipnee replied:
I think the `instanceof ...` exceptions are also non-retriable, and I think they need to be handled first.

philipnee replied:
so the if else blocks becomes a bit fragmented.  or we could do:
```
if (!future.isRetriable()) {
  if ( ... instance of ... ) { continue; }
  throw ...
}

{rest of the logic there}
```

philipnee replied:
However, this is a bit more nested, which can be harder to read

philipnee committed:
Merge branch 'apache:trunk' into kafka-12639_exit_upon_expired_timer

guozhangwang commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1484,6 +1484,8 @@ public void testRebalanceWithMetadataChange() {
                 Utils.mkMap(Utils.mkEntry(topic1, 1), Utils.mkEntry(topic2, 1))));
         client.respond(joinGroupFollowerResponse(1, consumerId, "leader", Errors.NOT_COORDINATOR));
         client.prepareResponse(groupCoordinatorResponse(node, Errors.NONE));
+        coordinator.poll(time.timer(0)); // failing joinGroup request will require re-poll in order to retry
```
It's not very clear to me why here and line 3403 below we need additional polls since the test scenarios seems irrelevant to error cases?

philipnee replied:
good point... let me update the tests there. thanks

philipnee replied:
The NOT_COORDINATOR error originally should trigger retries; however, in the new code, it would exit due to an expired timer. Another way to do it is using poll(time.timer(1))

guozhangwang replied:
Ack, that makes sense.

guozhangwang reviewed (commented):
@philipnee thanks for the added unit test. I made another pass and left some more comments.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -3398,7 +3400,8 @@ public void testPrepareJoinAndRejoinAfterFailedRebalance() {
 
             // Join future should succeed but generation already cleared so result of join is false.
-            res = coordinator.joinGroupIfNeeded(time.timer(1));
+            coordinator.joinGroupIfNeeded(time.timer(0));
```
Similar here, the timer is expired upon IllegalGenerationException, the loop would continued in the original code, but now it would exit.  I guess we could try to poll for a bit longer, like 3ms instead of 0ms.

philipnee committed:
Use a timer in the test as the loop will exit upon expired timer

philipnee committed:
Documentation enhancement

philipnee commented:
Here are a couple things I updated:
1. Added some documentation to clarify the intent, but I didn't rewrite it as nested if "can be" harder to read.
2.  Added non zero timeouts for the tests as our timer now is stricter and will explicitly exit upon expiration.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -500,14 +500,24 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     requestRejoin(shortReason, fullReason);
                 }
 
+                // 4 special non-retriable exceptions that we want to retry, as long as the timer hasn't expired.
```
I think here the comment is not to just state what the code did, since readers can just understand that from the code :P instead what we want to emphasize is to remind future contributors that they should be careful to not change the precedence ordering of this logic unnecessarily.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -500,14 +500,24 @@ boolean joinGroupIfNeeded(final Timer timer) {
-                    exception instanceof IllegalGenerationException ||
-                    exception instanceof RebalanceInProgressException ||
-                    exception instanceof MemberIdRequiredException)
+                        exception instanceof IllegalGenerationException ||
```
Ah thanks for the clarifications!

Thinking about this a bit more (sorry for getting back and forth..), I now concerned a bit more that for some usage patterns where `poll` call would be triggered less frequently, we may not be coming back to handle these four exceptions while at the same time the broker is ticking and waiting for the join-group request to be re-sent. Hence I'm changing my mind to lean a bit more to honor the exception types for immediate handling than the timeouts --- again, sorry for going back and forth...

So I think we would define the ordering as the following:

1. For un-retriable exception, always try to handle immediately and not honor the timer.
2. Otherwise, honor the timer.

In that case, we could just go back to the first time you made the change, i.e. just add the 

```
if (timer.isExpired())
                        return false;
```

After the `if/else-if` block. Still it's better to comment that above ordering is diligently designed as such.

philipnee replied:
Hey thanks for the comments again and absolutely no apology is needed there! I guess, as we all know, rebalancing is full of subtleties, so it makes sense to be careful about these non-retriable exception case. I think it's a good idea to keep the original behavior consistent, in case of unexpected breakage. Updating the PR.

philipnee committed:
Revert back from previous changes

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1484,7 +1484,8 @@ public void testRebalanceWithMetadataChange() {
         client.prepareResponse(groupCoordinatorResponse(node, Errors.NONE));
-        coordinator.poll(time.timer(0));
+        assertFalse(client.hasInFlightRequests());
+        coordinator.poll(time.timer(1));
```
note: we need to add a timeout here to give the retry a second chance, because in the new code, the timer is checked and causes the method to exit.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -501,13 +501,18 @@ boolean joinGroupIfNeeded(final Timer timer) {
+                        exception instanceof RebalanceInProgressException ||
+                        exception instanceof MemberIdRequiredException)
                     continue;
                 else if (!future.isRetriable())
```
the previous logic was reverted with some autocorrection to the indentation.

philipnee committed:
documentation

guozhangwang reviewed (approved):
LGTM, waiting for the jenkins job to complete.

philipnee commented:
Hmm. I think these tests are flaky actually
```
Build / JDK 17 and Scala 2.13 / shouldPauseStandbyTaskAndNotTransitToUpdateStandbyAgain() – org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest
30s
Build / JDK 17 and Scala 2.13 / shouldPauseActiveTaskAndTransitToUpdateStandby() – org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest
30s
Build / JDK 17 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
2m 0s
Build / JDK 11 and Scala 2.13 / testListenerConnectionRateLimitWhenActualRateAboveLimit() – kafka.network.ConnectionQuotasTest
19s
Build / JDK 11 and Scala 2.13 / shouldRemovePausedAndUpdatingTasksOnShutdown() – org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest
30s
Build / JDK 11 and Scala 2.13 / shouldPauseStandbyTaskAndNotTransitToUpdateStandbyAgain() – org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest
31s
```

guozhangwang commented:
The test failures are not relevant (but some of them are related to DefaultStateUpdaterTest.. sigh).

guozhangwang commented:
Merged to trunk.

philipnee commented:
yeah the DefaultStateUpdaterTest has been failing from time to time... not sure why 😭
